tools/c7n_azure/c7n_azure/resources/app_gateway.py

```diff
@@ -33,3 +36,56 @@ class resource_type(ArmResourceManager.resource_type):
             'resourceGroup'
         )
         resource_type = 'Microsoft.Network/applicationGateways'
+
+
+@ApplicationGateway.filter_registry.register('waf')
+class ApplicationGatewayWafFilter(Filter):
+    """
+    Filter Application Gateways using WAF rule configuration
+
+    :example:
+
+    Return all the App Gateways which have rule '944240' disabled.
+
+    .. code-block:: yaml
+
+        policies:
+          - name: test-app-gateway
+            resource: azure.application-gateway
+            filters:
+              - type: waf
+                override_rule: 944240
+                state: disabled
+    """
+    
+    schema = type_schema(
+        'waf',
+        required=['override_rule', 'state'],
+        **{
+            'override_rule': {'type': 'number'},
+            'state': {'type': 'string', 'enum': ['disabled']}}
+    )
+
+    def process(self, resources, event=None):
+
+        filter_override_rule = self.data.get('override_rule')
+        filter_state = self.data.get('state')
+
+        client = self.manager.get_client()
+        result = []
+
+        for resource in resources:
+            if 'firewallPolicy' in resource['properties']:
+                waf_policy_name = resource['properties']['firewallPolicy']['id'].split('/')[-1]
+
+                app_gate_waf = client.web_application_firewall_policies.\
+                    get(resource['resourceGroup'],waf_policy_name)
+
+                for rule_set in app_gate_waf.managed_rules.managed_rule_sets:
+                    for group in rule_set.rule_group_overrides:
+                        for rule in group.rules:
+                            if filter_override_rule == int(rule.rule_id) \
+                                and filter_state.lower() == rule.state.lower():
+                                result.append(resource)
```

Comment:
this is potentially going to return duplicate resources in the result, a filter is passed a unique set of resources, it shouldn't return a list with duplicates.

Reply:
it's iterating through the list of unique app_gateway resources in line 77. And each app_gateway can be associated to the one WAF. So it will only add app_gateway to the result set if the filter condition is true. It will not return the duplicate set of the app_gateways in result set.